### PR TITLE
[heft] Add TypeScript build support to Heft.

### PR DESCRIPTION
**Pull request by iclanton**

### CODE REVIEW

File: apps/heft/src/cli/actions/BuildAction.ts [hunk 1]
```diff
@@ -74,17 +74,69 @@ export interface ICopyStaticAssetsConfiguration extends ISharedCopyStaticAssetsC
...
+  additionalModuleKindsToEmit?: IEmitModuleKind[] | undefined;
+}
+
+/**
+ * @public
+ */
+export interface ITypescriptConfiguration extends ISharedTypescriptConfiguration {
+  tsconfigPaths: string[];
+  tslintConfigPath: string | undefined;
+  isLintingEnabled: boolean | undefined;
```
Comment (octogonz) [inline apps/heft/src/cli/actions/BuildAction.ts:121]: How will TSLint vs ESLint be handled with this design?

File: apps/heft/src/cli/actions/BuildAction.ts [hunk 2]
```diff
@@ -74,17 +74,69 @@ export interface ICopyStaticAssetsConfiguration extends ISharedCopyStaticAssetsC
...
+   */
+  additionalModuleKindsToEmit?: IEmitModuleKind[] | undefined;
+}
+
+/**
+ * @public
+ */
+export interface ITypescriptConfiguration extends ISharedTypescriptConfiguration {
+  tsconfigPaths: string[];
+  tslintConfigPath: string | undefined;
```
Comment (octogonz) [inline apps/heft/src/cli/actions/BuildAction.ts:120]: If the path is configurable, how will the VS Code extension find it?

**Comment by octogonz** [comment]
Now that we have a workable toolchain, we need to add some test projects under the `build-tests` folder. Otherwise there's no way to test these features.

File: apps/heft/src/cli/actions/BuildAction.ts (hunk 2 above)
Comment (iclanton) [reply apps/heft/src/cli/actions/BuildAction.ts:120]: Hmm that's a good point. Perhaps we shouldn't bother making it configurable.

File: apps/heft/src/plugins/PackageJsonConfigurationPlugin.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    });
+  }
+
+  private _updateTypescriptConfiguration(
+    projectPackageJson: IPackageJson,
+    typescriptConfiguration: ITypescriptConfiguration
+  ): void {
+    if (projectPackageJson.private !== true) {
+      // Copy if the package is intended to be published
+      typescriptConfiguration.copyFromCacheMode = 'copy';
```
Comment (octogonz) [inline apps/heft/src/plugins/PackageJsonConfigurationPlugin.ts:31]: It is somewhat counterintuitive that the `"private": true` field in **package.json** would influence the compiler. (It's often not even accurate in a Rush repo.)

Suggestions:
1. When this heuristic is applied, write something to the logs like 
   ```
   Defaulting to copyFromCacheMode="copy" because private=true in package.json
   ``` 
2. Make this a default that can be overridden easily via a simple config file (without making a rig)

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,1029 @@
...
+  IResolveModuleNameResolutionHost
+} from './privateTypings/ExtendedTypescript';
+
+import { SubprocessRunnerBase } from '../../utilities/subprocess/SubprocessRunnerBase';
+import { Async } from '../../utilities/Async';
+import { resolvePackagePath } from '../../utilities/resolvePackage';
+import { IExtendedLinter } from './privateTypings/ExtendedTslint';
+import { IEmitModuleKindBase, ISharedTypescriptConfiguration } from '../../cli/actions/BuildAction';
+
+const ASYNC_LIMIT: number = 100;
```
Comment (octogonz) [inline apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:31]: This should be configurable. A default of `100` seems very high:
- A hard drive cannot efficiently write 100 files simultaneously
- In a Rush repo with `--parallelism` equal to the number of CPU cores, the task switching overhead would be pretty inefficient for 100 concurrent operations
- On MacOS the  maximum number of file handles can be relatively low, see https://github.com/microsoft/rushstack/issues/1934#issuecomment-648489001

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts [hunk 5]
```diff
@@ -0,0 +1,1029 @@
...
+  tslintPath: string | undefined;
+
+  /**
+   * The path of project's build cache folder
+   */
+  buildCacheFolder: string;
+}
+
+interface ICachedEmitModuleKind<TModuleKind> extends IEmitModuleKindBase<TModuleKind> {
+  cacheOutFolderPath: string;
```
Comment (octogonz) [inline apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:60]: The purpose of these interfaces is not obvious.  It would be good to document them.
For each setting:
- What does it do?
- If it's a string, show an example value
- If it's a path, what is it relative to?
- If it's `undefined` then what is the behavior?

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts [hunk 6]
```diff
@@ -0,0 +1,1029 @@
...
+}
+
+interface IFileToWrite {
+  filePath: string;
+  data: string;
+}
+
+interface ITsLintCacheData {
+  cacheVersion: string;
+  fileVersions: [string, string][];
```
Comment (octogonz) [inline apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:71]: Use a proper interface instead of `[string, string]`

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts [hunk 7]
```diff
@@ -0,0 +1,1029 @@
...
+
+interface IRunTslintOptions {
+  tslint: typeof Tslint;
+  tsProgram: IExtendedProgram;
+  typescriptFilenames: Set<string>;
+  measurePerformance: PerformanceMeasurer;
+  changedFiles: Set<IExtendedSourceFile>;
+}
+
+type PerformanceMeasurer = <TResult extends object | void>(
```
Comment (octogonz) [inline apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:82]: This seems fairly general -- why is it defined inside TypescriptBuilder.ts?

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts [hunk 8]
```diff
@@ -0,0 +1,1029 @@
...
+import {
+  ExtendedTypescript,
+  IEmitResolver,
+  IExtendedEmitResult,
+  IEmitHost,
+  IEmitTransformers,
+  IExtendedProgram,
+  IExtendedSourceFile,
+  IResolveModuleNameResolutionHost
+} from './privateTypings/ExtendedTypescript';
```
Comment (octogonz) [inline apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:23]: `TypeScript` is capitalized incorrectly everywhere in this PR.

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts [hunk 9]
```diff
@@ -0,0 +1,1029 @@
...
+        ...tsProgram.getSyntacticDiagnostics(),
+        ...tsProgram.getGlobalDiagnostics(),
+        ...tsProgram.getSemanticDiagnostics()
+      ];
+      const _diagnostics: ReadonlyArray<Typescript.Diagnostic> = ts.sortAndDeduplicateDiagnostics(
+        rawDiagnostics
+      );
+      return { diagnostics: _diagnostics };
+    });
+    this._terminal.writeVerboseLine(`Diagnostics: ${diagnosticsDurationMs}ms`);
```
Comment (octogonz) [inline apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:291]: FYI the individual diagnostic times can be interesting. For example, after `git pull` from master, we have a fairly strong assurance that everything compiles correctly. Thus we could greatly accelerate a local `rush install && rush build` by optionally skipping type checking, i.e. telling the compiler to only do the bare minimum analysis needed for transpilation.

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts [hunk 10]
```diff
@@ -0,0 +1,1029 @@
...
+  }
+
+  public async _runBuild(
+    ts: ExtendedTypescript,
+    tslint: typeof Tslint | undefined,
+    measureTsPerformance: PerformanceMeasurer,
+    measureTsPerformanceAsync: PerformanceMeasurerAsync
+  ): Promise<void> {
+    // Ensure the cache folder exists
+    this._fileSystem.ensureFolder(this._configuration.buildCacheFolder);
```
Comment (octogonz) [inline apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:234]: Is there an easy way to clear the cache?

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts (hunk 5 above)
Comment (octogonz) [reply apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:60]: The TypescriptBuilder.ts is likely to be a primary focus for optimizations, so it would be worthwhile to make the code as easy for people to understand as possible.

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts [hunk 11]
```diff
@@ -0,0 +1,1029 @@
...
+      terminalMessage = ['  ', ...terminalMessage];
+    }
+
+    switch (diagnostic.category) {
+      case ts.DiagnosticCategory.Error: {
+        this._terminal.writeErrorLine(...terminalMessage);
+        break;
+      }
+
+      case ts.DiagnosticCategory.Warning: {
```
Comment (octogonz) [inline apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:531]: A major annoyance with our current toolchain is that it does not distinguish "errors" from "warnings" very well (for both ESLint and TypeScript).

Consider this input:
```ts
function example(): void {
  let x = new ForgottenImport();
  let p = new Promise(() => { });
}
```
Heft's output looks like this:
```
[typescript] Encountered 3 TypeScript errors:
[typescript]   ERROR: src\index.ts(5,7) - (TS6133) 'x' is declared but its value is never read.
[typescript]   ERROR: src\index.ts(5,15) - (TS2304) Cannot find name 'ForgottenImport'.
[typescript]   ERROR: src\index.ts(6,7) - (TS6133) 'p' is declared but its value is never read.
[typescript] Encountered 5 TSLint errors:
[typescript]   ERROR: src\index.ts(6,29) - (no-empty) block is empty
[typescript]   ERROR: src\index.ts(5,7) - (prefer-const) Identifier 'x' is never reassigned; use 'const' instead of 'let'.
[typescript]   ERROR: src\index.ts(6,7) - (prefer-const) Identifier 'p' is never reassigned; use 'const' instead of 'let'.
[typescript]   ERROR: src\index.ts(5,7) - (typedef) expected variable-declaration: 'x' to have a typedef
[typescript]   ERROR: src\index.ts(6,7) - (typedef) expected variable-declaration: 'p' to have a typedef
```

This error is a serious mistake, such that the code is unlikely to work at all:
```
[typescript]   ERROR: src\index.ts(5,15) - (TS2304) Cannot find name 'ForgottenImport'.
```

But it is buried in a haystack of other 5 alarm "ERROR" messages complaining about absolutely trivial details, that are totally normal while a person is debugging the code.  Arguably those should really be modeled as "warnings".

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts [hunk 12]
```diff
@@ -0,0 +1,1029 @@
...
+      );
+      for (const diagnostic of diagnostics) {
+        this._printDiagnosticMessage(ts, diagnostic, true);
+      }
+
+      compilationFailed = true;
+    }
+
+    if (tslintResult?.failures?.length) {
+      this._terminal.writeWarningLine(
```
Comment (octogonz) [inline apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:470]: Continuing my example from above, the log output is colored as follows:

In red color:
```
[typescript] Encountered 3 TypeScript errors:
[typescript]   ERROR: src\index.ts(5,7) - (TS6133) 'x' is declared but its value is never read.
[typescript]   ERROR: src\index.ts(5,15) - (TS2304) Cannot find name 'ForgottenImport'.
[typescript]   ERROR: src\index.ts(6,7) - (TS6133) 'p' is declared but its value is never read.
```

In yellow color:
```
[typescript] Encountered 5 TSLint errors:
[typescript]   ERROR: src\index.ts(6,29) - (no-empty) block is empty
[typescript]   ERROR: src\index.ts(5,7) - (prefer-const) Identifier 'x' is never reassigned; use 'const' instead of 'let'.
[typescript]   ERROR: src\index.ts(6,7) - (prefer-const) Identifier 'p' is never reassigned; use 'const' instead of 'let'.
[typescript]   ERROR: src\index.ts(5,7) - (typedef) expected variable-declaration: 'x' to have a typedef
[typescript]   ERROR: src\index.ts(6,7) - (typedef) expected variable-declaration: 'p' to have a typedef
```

Questions:
- Why are warnings referred to as "errors"?
- Why are TSLint problems prefixed with `[typescript]`?

There are probably nerdy excuses, but as a user experience it seems not right.

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts [hunk 13]
```diff
@@ -0,0 +1,1029 @@
...
+
+    switch (diagnostic.category) {
+      case ts.DiagnosticCategory.Error: {
+        this._terminal.writeErrorLine(...terminalMessage);
+        break;
+      }
+
+      case ts.DiagnosticCategory.Warning: {
+        this._terminal.writeWarningLine(...terminalMessage);
+        break;
```
Comment (octogonz) [inline apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:533]: Also, continuing the above example log output, the most important information should appear towards the end of the log (i.e. the part that doesn't scroll out of view).  So maybe these details should be moved much earlier (and maybe squashed onto a single line?):

```
Heft version: 0.1.0
Node version: v12.17.0
```

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts [hunk 14]
```diff
@@ -0,0 +1,1029 @@
...
+    // Code taken from reference example
+    if (diagnostic.file) {
+      const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start!);
+      const message: string = ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n');
+      const buildFolderRelativeFilename: string = path.relative(
+        this._configuration.buildFolder,
+        diagnostic.file.fileName
+      );
+      terminalMessage = [
+        Colors.red(`ERROR: ${buildFolderRelativeFilename}(${line + 1},${character + 1})`),
```
Comment (octogonz) [inline apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:511]: Do we still prefer the parenthesized notation?
```
src/folder/File.ts(12,34)
```
`tsc` now defaults to "Unix" notation, which I think is supported by a wider range of editors:
```
src/folder/File.ts:12:34
```

This [ESLint page](https://eslint.org/docs/user-guide/formatters/) lists all the popular formats.  (The one we're using is called "visualstudio".)

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts [hunk 15]
```diff
@@ -0,0 +1,1029 @@
...
+          compilerHost: _compilerHost
+        };
+      }
+    );
+    this._terminal.writeVerboseLine(`Configure: ${configureDurationMs}ms`);
+    //#endregion
+
+    this._validateTsconfig(ts, tsconfig);
+
+    ts.createWatchProgram(compilerHost);
```
Comment (octogonz) [inline apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:220]: Is there a way to watch for changes in `tsconfig.json`?  Currently that's ignored.

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts [hunk 16]
```diff
@@ -0,0 +1,1029 @@
...
+      )
+    );
+    //#endregion
+
+    const typescriptFilenames: Set<string> = new Set(tsconfig.fileNames);
+
+    //#region TSLINT
+    let tslintResult: Tslint.LintResult | undefined = undefined;
+    if (tslint) {
+      tslintResult = await this._runTslintAsync({
```
Comment (octogonz) [inline apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:328]: This file is getting kind of large. Would it be easy to split `_runTslintAsync()` into a separate file? That way ESLint and TSLint won't both be mixed up with the other unrelated code in TypeScriptBuilder.ts

File: apps/heft/src/plugins/TypescriptPlugin/privateTypings/ExtendedTypescript.ts [hunk 17]
```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
```
Comment (octogonz) [inline apps/heft/src/plugins/TypescriptPlugin/privateTypings/ExtendedTypescript.ts:2]: API Extractor calls this file [TypeScriptInternals.ts](https://github.com/microsoft/rushstack/blob/a30cdf5b3caed87c63ee55f2c91dc1409fa7d29b/apps/api-extractor/src/analyzer/TypeScriptInternals.ts#L16).  

I think that's a more intuitive name than ExtendedTypeScript.ts, if this file is really just a bunch of compiler internals.

(BTW I don't see much value in sharing compiler internals declarations between API Extractor and Heft)

File: apps/heft/src/schemas/typescript.schema.json [hunk 18]
```diff
@@ -0,0 +1,46 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "title": "TypeScript Build Configuration",
+  "description": "Defines optional options for TypeScript build.",
```
Comment (octogonz) [inline apps/heft/src/schemas/typescript.schema.json:4]: Is this a declaration of `tsconfig.json`? Or is it proprietary to Heft? We should document that more clearly.

Also, how do I use it?  `heft-test-01` doesn't seem to have any config file conforming to this schema.

File: apps/heft/src/schemas/typescript.schema.json [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+          "outFolderPath": {
+            "type": "string",
+            "pattern": "[^\\\\\\/]"
+          }
+        },
+        "required": ["moduleKind", "outFolderPath"]
+      }
+    },
+
+    "disableTslint": {
```
Comment (octogonz) [inline apps/heft/src/schemas/typescript.schema.json:41]: TSLint was deprecated over a year ago, and the whole world has moved on.  It really shouldn't be "on-by-default".

File: apps/heft/src/utilities/Constants.ts [hunk 20]
```diff
@@ -3,4 +3,6 @@
 
 export class Constants {
   public static projectHeftFolderName: string = '.heft';
+
+  public static buildCacheFolderName: string = 'build-cache';
```
Comment (octogonz) [inline apps/heft/src/utilities/Constants.ts:7]: Document the path where this folder can be found. And what it's for. Seems like a somewhat important specification.

File: apps/heft/src/utilities/fileSystem/IExtendedFileSystem.ts [hunk 21]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { FileSystem, IFileSystemCreateLinkOptions } from '@rushstack/node-core-library';
+
+export type FilesystemType = Omit<typeof FileSystem, 'prototype'>;
```
Comment (octogonz) [inline apps/heft/src/utilities/fileSystem/IExtendedFileSystem.ts:6]: There are two kinds of API interfaces:

- **easy-to-use**: A rich set of different operations for various usage scenarios, with lots of handy options
- **easy-to-implement**: A minimal set of "axioms" that fulfill the core functions needed to support higher-level systems

The design requirements are diametrically opposite for these two cases, so it's common to find them connected by an adapter that maps all the **easy-to-use** APIs onto a barebones plugin layer.

The `FileSystem` signature is designed to be **easy-to-use**, whereas your `CachedFileSystem` seems like it needs an `easy-to-implement` contract.  Otherwise every little change we make to `FileSystem` will break Heft and require special fixups to adapt it for caching (if I understand correctly).  Also it is telling that:

- `IExtendedFileSystem` has to add extra operations like `readFolderFilesAndDirectories()` that could not be cached reasonably
- Quite a lot of the `IExtendedFileSystem` are never actually used

For a good example of an **easy-to-implement** abstraction, take a look at the compiler's `ts.System` (which defines the bare minimum set of OS operations required to port the compiler to a new JavaScript runtime).  

Fixing this might be out of scope for this PR, but it seems like `CachedFileSystem` is going to run into architectural problems with its current design. And cache bugs are not fun bugs. :-)

File: apps/heft/src/utilities/resolvePackage.ts [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+ */
+export interface IResolvePackageOptions {
+  doNotResolveSymlinks: boolean;
+}
+
+/**
+ * Gets the path of a given package name
+ * @public
+ */
+export function resolvePackage(
```
Comment (octogonz) [inline apps/heft/src/utilities/resolvePackage.ts:22]: This is exported, but nothing seems to use it.  Also, it's marked as `@public` but not actually exported (nor should it be, since functions do not make good public APIs).

File: apps/heft/src/utilities/resolvePackage.ts [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+    } catch (e2) {
+      throw e1;
+    }
+  }
+}
+
+/**
+ * Resolves a path in a package, relative to another path.
+ */
+export function resolvePackagePath(
```
Comment (octogonz) [inline apps/heft/src/utilities/resolvePackage.ts:63]: Why aren't we using the [resolve](https://www.npmjs.com/package/resolve) library? This seems like a complex wheel to reinvent. (?)

File: apps/heft/src/utilities/subprocess/SubprocessRunnerBase.ts [hunk 24]
```diff
@@ -0,0 +1,161 @@
...
+export const SUBPROCESS_RUNNER_CLASS_LABEL: unique symbol = Symbol('IsSubprocessModule');
+export const SUBPROCESS_RUNNER_INNER_INVOKE: unique symbol = Symbol('SubprocessInnerInvoke');
+
+interface ISubprocessExitMessage extends ISubprocessMessage {
+  type: 'exit';
+  errorMessage?: string;
+  errorStack?: string;
+}
+
+export abstract class SubprocessRunnerBase<TSubprocessConfiguration> {
```
Comment (octogonz) [inline apps/heft/src/utilities/subprocess/SubprocessRunnerBase.ts:29]: This needs some docs. It's a fairly complex system with a proprietary protocol, but from looking at the code it's not immediately apparent what it is even used for.

File: common/reviews/api/heft.api.md [hunk 25]
```diff
@@ -14,6 +14,7 @@ import { IPackageJson } from '@rushstack/node-core-library';
 import { ITerminalProvider } from '@rushstack/node-core-library';
 import { SyncHook } from 'tapable';
 import { Terminal } from '@rushstack/node-core-library';
+import * as TRushStackCompiler from '@microsoft/rush-stack-compiler-3.7';
```
Comment (octogonz) [inline common/reviews/api/heft.api.md:17]: @D4N14L was able to eliminate this dependency for RSC. Can we also eliminate it for Heft? 

It is an expensive download for a **package.json** dependency, and it seems that we only use it for typing info.

File: apps/heft/src/schemas/typescript.schema.json (hunk 19 above)
Comment (iclanton) [reply apps/heft/src/schemas/typescript.schema.json:41]: It's only "on" if a `tslint.json` file is present. This option only exists to disable it iff that file exists.

File: apps/heft/src/utilities/resolvePackage.ts (hunk 23 above)
Comment (iclanton) [reply apps/heft/src/utilities/resolvePackage.ts:63]: This does use the `resolve` library. See line 4.

File: apps/heft/src/schemas/typescript.schema.json (hunk 18 above)
Comment (iclanton) [reply apps/heft/src/schemas/typescript.schema.json:4]: No, this file provides some heft-specific options. I'll add usage of it to `heft-test-01`

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts (hunk 15 above)
Comment (iclanton) [reply apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:220]: I think that'll be pretty difficult to handle. Maybe that's a feature we can add later.

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts (hunk 6 above)
Comment (iclanton) [reply apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:71]: TSLint gives up a `Map<string, string>` and we can use `Array.from(...)` on it to get a `[string, string][]`. Changing this to an object will make the cache file larger and require looping through the `Map` entries to construct an array of objects, which is generally pointless work for a file that's never expected to be human-read.

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts (hunk 10 above)
Comment (iclanton) [reply apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:234]: Deleting the cache folder is the only way to clear it right now. I can add a flag to the `clean` verb to also delete the cache.

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts (hunk 6 above)
Comment (octogonz) [reply apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:71]: Can you document the data type then and give an example value?

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts (hunk 13 above)
Comment (iclanton) [reply apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:533]: There isn't really a more natural place to log this. I think we should spend some time designing the whole logging system, and moving this around can be part of that.

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts (hunk 12 above)
Comment (iclanton) [reply apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:470]: I've given the tslint logging messages their own prefix for now. I think we need to have a longer design discussion about this.

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts (hunk 16 above)
Comment (iclanton) [reply apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:328]: Hmm that'll be kind of a PITA. Let's consider refactoring when ESLint support is added.

File: apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts (hunk 14 above)
Comment (iclanton) [reply apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts:511]: I don't have a strong opinion. I can change it.

File: apps/heft/src/cli/actions/BuildAction.ts (hunk 1 above)
Comment (iclanton) [reply apps/heft/src/cli/actions/BuildAction.ts:121]: It'll really depend on if ESLint can take an existing TypeScript analysis or not. If it can, it should live parallel to TSLint. If it can't, there's no reason to marry it to the TypeScript builder.

There's also no technical reason a project can't use both TSLint and ESLint. I can't think of a reason anyone would want to do that, but it's totally technically possible.

File: apps/heft/src/utilities/fileSystem/IExtendedFileSystem.ts (hunk 21 above)
Comment (iclanton) [reply apps/heft/src/utilities/fileSystem/IExtendedFileSystem.ts:6]: Yeah I don't like this design. We should redesign this, but let's do that in a separate PR.

File: apps/heft/src/cli/actions/BuildAction.ts (hunk 1 above)
Comment (octogonz) [reply apps/heft/src/cli/actions/BuildAction.ts:126]: @iclanton  We actually have some projects at work that invoke ESLint and TSLint together, because the person who set it up didn't think it through, and it never got fixed. It's a confusing experience. Maybe it could be allowed to facilitate a migration, but it shouldn't be easy to do by mistake.

Comment (octogonz) [reply apps/heft/src/cli/actions/BuildAction.ts:126]: > It'll really depend on if ESLint can take an existing TypeScript analysis or not.

This is an interesting design question.

**Pro:**
- I don't think ESLint supports this today, but the [typescript-eslint project](https://github.com/typescript-eslint/typescript-eslint/) is maintained by a different group that tend to be quite fast with turning around PRs (since it is still relatively new), so we could realistically contribute this feature.
- The `ts.Program` [gets created here](https://github.com/typescript-eslint/typescript-eslint/blob/7c7032322f55d9492e21d3bfa5da16da1f05cbce/packages/typescript-estree/src/create-program/createDefaultProgram.ts#L20), and the semantic analysis [happens like this](https://github.com/typescript-eslint/typescript-eslint/blob/3a15413d87a3429ebf19af2cc5db76c9e7ffe4e7/packages/typescript-estree/src/semantic-or-syntactic-errors.ts#L28), which would be pretty easy to hook into. Also, the scenario we care about optimizing invokes ESLint exactly once, without any watch or updates. So the work involved seems not expensive.

**Con:**
- On the other hand, in a Rush repo the big win for ESLint would be the multiphase feature, where linting would happen in the background as a later phase that does not block downstream projects. With this optimization, the actual CPU cost of linting might matter less.
